.github/workflows/ci.yml

```
@@ -26,67 +28,171 @@ jobs:
         with:
           go-version: "1.24"
 
-      - name: Install dependencies
+      - name: Install build dependencies (for linting/testing on host)
         run: |
           sudo apt-get update
-          sudo apt-get install -y clang llvm make libbpf-dev
+          sudo apt-get install -y clang clang-format llvm make libbpf-dev build-essential
 
-      - name: Install Go tools
+      - name: Install Go development tools (golangci-lint, bpf2go)
         run: |
           go install github.com/cilium/ebpf/cmd/bpf2go@latest
+          go install github.com/golangci/golangci-lint/cmd/golangci-lint@latest
 
-      - name: Download dependencies
+      - name: Download Go modules
         run: go mod download
 
-      - name: Generate eBPF bindings
+      - name: Run linters
+        run: make lint
+
+      # Generate eBPF Go bindings on the host (amd64) for local test builds.
+      # This doesn't produce cross-arch binaries for release, just the Go source files.
+      - name: Generate eBPF Go bindings
         run: make generate
 
-      - name: Build
-        run: make build
+      - name: Build local binary for testing
+        run: |
+          mkdir -p build
+          # Ensure CGO_ENABLED=1 for local test build as eBPF needs it
+          GOOS=linux GOARCH=amd64 CGO_ENABLED=1 go build -o build/mcpspy-linux-amd64 ./cmd/mcpspy
 
       - name: Run unit tests
         run: make test
 
       - name: Run end-to-end tests
-        run: make test-e2e
+        # Requires root/privileged access due to eBPF.
+        run: sudo -E make test-e2e
+
+# Job to build cross-platform binaries as release assets
+  build_binaries:
+    needs: [test]
```

This is needed only for release

```suggestion
    if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags/')
    needs: [test]
```

---

The current CI pipeline builds a Docker image twice in [docker, release] jobs: once with the latest tag and again with the vX.X.X tag in release. This approach will build a Docker image only once, but add all required tags simultaneously.

---

I agree.
but still, the `build_binaries` job isn't related to pull requests, and should happen on pushes / tags. as you build it for the docker images which are pushed later only with the `push` event.

---

Good catch! Should be fixed now